### PR TITLE
Create the protocol generator

### PR DESCRIPTION
**Pull request by Kapin**
Generator that uses the database to create xml string output. Includes a lot of tests. I will create the routes, and other bits later.

Tests need a lot of future cleanup as there is a ridiculous amount of redundant code

### CODE REVIEW

File: src-backend/api/generator.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        props = self._get_properties()
+        return ElementTree.SubElement(parent, self.name, props)
+
+
+class ProtocolBuilder:
+    @classmethod
+    def _prettify(cls, elem):
+        rough_string = ElementTree.tostring(elem, 'utf-8')
+        reparsed = minidom.parseString(rough_string)
+        return reparsed.toprettyxml(indent="  ")
```
Comment (Trinovantes) [inline src-backend/api/generator.py]: 4 spaces

File: src-backend/api/generator.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def _generateTree(cls, owner, pk):
+        try:
+            procedure = Procedure.objects.get(pk=pk)
+        except Procedure.DoesNotExist:
+            raise ValueError('Invalid procedure id')
+
+        if procedure.owner != owner:
+            raise ValueError('Invalid owner')
+
+        procedureElement = ProcedureGenerator(procedure).generate()
```
Comment (Trinovantes) [inline src-backend/api/generator.py]: snake case

File: src-backend/api/tests/test_generator.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        test_user.save()
+
+        self.procedure = Procedure.objects.create(
+            author='tester',
+            title='test procedure',
+            owner=test_user
+        )
+
+        self.generator = ProcedureGenerator(self.procedure)
+        self.procedureElement = self.generator.generate()
```
Comment (Trinovantes) [inline src-backend/api/tests/test_generator.py]: snake case

File: src-backend/api/generator.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return props
+
+    def generate(self, parent):
+        props = self._get_properties()
+        return ElementTree.SubElement(parent, self.name, props)
+
+
+class ProtocolBuilder:
+    @classmethod
+    def _prettify(cls, elem):
```
Comment (Trinovantes) [inline src-backend/api/generator.py]: if these are supposed to be private, should be prefixed with `__`

File: src-backend/api/tests/test_generator.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            concept='COMPLAINT PICTURE',
+            question='Add a Picture:',
+            answer='',
+            page=page4
+        )
+
+        self.procedure.save()
+
+    def test_generates_tree(self):
+        tree = ProtocolBuilder._generateTree(self.test_user, self.procedure.id)
```
Comment (Trinovantes) [inline src-backend/api/tests/test_generator.py]: calling private method?

**Comment by Trinovantes** [comment]
Resolves #48

File: src-backend/api/generator.py (hunk 1 above)
Comment (Kapin) [inline src-backend/api/generator.py]: will fix

File: src-backend/api/generator.py (hunk 4 above)
Comment (Kapin) [inline src-backend/api/generator.py]: http://stackoverflow.com/questions/1301346/the-meaning-of-a-single-and-a-double-underscore-before-an-object-name-in-python

Seems to indicate this is the correct convention.

File: src-backend/api/tests/test_generator.py (hunk 5 above)
Comment (Kapin) [inline src-backend/api/tests/test_generator.py]: It's a test. There is no concept of private in Python, it's only a thing saying "don't touch this".

File: src-backend/api/generator.py (hunk 4 above)
Comment (Trinovantes) [inline src-backend/api/generator.py]: I don't see why we shouldn't use `__` to be strict. If this is supposed to be conceptually private, it shouldn't be easily called by name.

File: src-backend/api/tests/test_generator.py (hunk 5 above)
Comment (Trinovantes) [inline src-backend/api/tests/test_generator.py]: I think this method should either be public or we only test the pre/post conditions of any public methods that will call this internally (http://c2.com/cgi/wiki?InappropriateIntimacy)

Comment (Kapin) [inline src-backend/api/tests/test_generator.py]: I have no issues making it public. Will change

File: src-backend/api/generator.py (hunk 4 above)
Comment (Kapin) [inline src-backend/api/generator.py]: Fair point.

**Comment by Trinovantes** [comment]
+1

**Comment by Trinovantes** [comment]
Does this include a Django view endpoint that responds with the xml?

**Comment by Kapin** [comment]
@Trinovantes it does not. I have a "prototype" working on my machine, but it's not serializing the xml properly.

**Comment by Kapin** [comment]
Closing in favour of #134

**Pull request closed without merging**